### PR TITLE
[Editor] Make the stamp annotations alt text readable by either VO or NVDA (bug 1912001)

### PR DESCRIPTION
**Pull request by calixteman**

### CODE REVIEW

File: web/struct_tree_layer_builder.js [hunk 1]
```diff
@@ -74,21 +74,30 @@ const PDF_ROLE_TO_HTML_ROLE = {
 const HEADING_PATTERN = /^H(\d+)$/;
 
 class StructTreeLayerBuilder {
-  #treeDom = undefined;
```
Comment (Snuffleupagus) [inline web/struct_tree_layer_builder.js:77]: Why was this changed?
Please revert this, see also comment below.

File: web/struct_tree_layer_builder.js [hunk 2]
```diff
@@ -74,21 +74,30 @@ const PDF_ROLE_TO_HTML_ROLE = {
...
+  #elementAttributes = new Map();
+
+  constructor(pdfPage) {
+    this.#promise = pdfPage.getStructTree();
   }
 
-  render(structTree) {
-    if (this.#treeDom !== undefined) {
+  async render() {
+    if (this.#treeDom) {
```
Comment (Snuffleupagus) [inline web/struct_tree_layer_builder.js:88]: This can lead to unnecessary re-parsing, please revert this.
Note how the `#walk`-method can return `null`, which the old condition thus handled correctly; see https://github.com/mozilla/pdf.js/blob/a41cd38387d09244c2a0c2405765efc8b3e9bce5/web/struct_tree_layer_builder.js#L120-L123

File: web/struct_tree_layer_builder.js [hunk 3]
```diff
@@ -104,7 +113,23 @@ class StructTreeLayerBuilder {
...
+      // in the annotation layer.
+      let added = false;
+      for (const child of structElement.children) {
+        if (child.type === "annotation") {
+          let attrs = this.#elementAttributes.get(child.id);
+          if (!attrs) {
+            attrs = Object.create(null);
+            this.#elementAttributes.set(child.id, attrs);
+          }
+          attrs["aria-label"] = removeNullCharacters(alt);
```
Comment (Snuffleupagus) [inline web/struct_tree_layer_builder.js:126]: Could multiple children have the same alt-text?
If so we can perhaps add a `const altText = removeNullCharacters(alt);` variable *above* the loop, and use that one both here in the the `setAttribute`-call below.

File: web/pdf_page_view.js [hunk 4]
```diff
@@ -1067,6 +1067,8 @@ class PDFPageView {
         showCanvas?.(true);
         await this.#finishRenderTask(renderTask);
 
+        this.structTreeLayer ||= new StructTreeLayerBuilder(this.pdfPage);
+
```
Comment (Snuffleupagus) [inline web/pdf_page_view.js:1071]: This seems like something that shouldn't be done unconditionally.
Also, why can't we use the local pdfPage-variable here?
```suggestion
        if (this.textLayer || this.annotationLayer) {
          this.structTreeLayer ||= new StructTreeLayerBuilder(pdfPage);
        }
```

File: web/pdf_page_view.js [hunk 5]
```diff
@@ -468,19 +472,15 @@ class PDFPageView {
     if (!this.textLayer) {
       return;
     }
-    this.structTreeLayer ||= new StructTreeLayerBuilder();
 
-    const tree = await (!this.structTreeLayer.renderingDone
-      ? this.pdfPage.getStructTree()
-      : null);
-    const treeDom = this.structTreeLayer?.render(tree);
+    const treeDom = await this.structTreeLayer.render();
```
Comment (Snuffleupagus) [inline web/pdf_page_view.js:476]: This is unfortunately not consistent with the old code, and may (likely) cause bugs in some edge-case!
Note how previously we'd only create and insert the structTree in the DOM *once*, however you've removed that logic here and I'm not at all convinced that it's safe/correct unfortunately.

File: web/pdf_page_view.js (hunk 4 above)
Comment (Snuffleupagus) [reply web/pdf_page_view.js:1071]: Also, how is overall performance affected since we'll now trigger structTree fetching/parsing *before* the textLayer is handled?

**Review by Snuffleupagus — CHANGES_REQUESTED** [review]
Based on a quick look, there's unfortunately a couple of bugs here.

File: web/pdf_page_view.js (hunk 4 above)
Comment (calixteman) [reply web/pdf_page_view.js:1071]: It shouldn't change the overall performances but just the scheduling of the tasks.
Without this patch, we've:
 - get the text for the text layer
 - while getting text the struct tree is retrieved
With this patch we've:
 - get the struct tree
 - get the text for the text layer

That said if you can imagine a way to have a perf regression, please tell me.

File: web/pdf_page_view.js [hunk 6]
```diff
@@ -468,19 +472,15 @@ class PDFPageView {
...
     }
-    this.structTreeLayer ||= new StructTreeLayerBuilder();
 
-    const tree = await (!this.structTreeLayer.renderingDone
-      ? this.pdfPage.getStructTree()
-      : null);
-    const treeDom = this.structTreeLayer?.render(tree);
-    if (treeDom) {
+    const treeDom = await this.structTreeLayer.render();
+    if (treeDom?.parentNode !== this.canvas) {
```
Comment (Snuffleupagus) [inline web/pdf_page_view.js:477]: Won't this attempt to insert a `null` structTree into the DOM, which I don't believe makes sense?
```suggestion
    if (treeDom && treeDom.parentNode !== this.canvas) {
```

File: web/struct_tree_layer_builder.js [hunk 7]
```diff
@@ -104,7 +114,24 @@ class StructTreeLayerBuilder {
...
-      htmlElement.setAttribute("aria-label", removeNullCharacters(alt));
+      // Don't add the label in the struct tree layer but on the annotation
+      // in the annotation layer.
+      let added = false;
+      const label = removeNullCharacters(alt);
+      for (const child of structElement.children) {
+        if (child.type === "annotation") {
+          let attrs = this.#elementAttributes.get(child.id);
+          if (!attrs) {
+            attrs = Object.create(null);
```
Comment (Snuffleupagus) [inline web/struct_tree_layer_builder.js:125]: Can't we make this a `Map` instead, so that it can be iterated "directly" without having to use `Object.entries(...)` in the annotationLayer-code above?
https://github.com/mozilla/pdf.js/pull/18658/files#diff-d7b040092cf6be09a276b785ba03c46e79bc8342cf4ae4fb1bc139dea8513425R3121

**Review by Snuffleupagus — COMMENTED** [review]
I've left a few more comments, but this is looking much better now.

Please run *all* tests once they are addressed, and assuming that everything works I'll do a final review.

File: web/pdf_page_view.js [hunk 8]
```diff
@@ -468,19 +472,15 @@ class PDFPageView {
     if (!this.textLayer) {
```
Comment (Snuffleupagus) [inline web/pdf_page_view.js:472]: Given the changes below, please let's ensure that the structTreeLayer is available when this method is invoked.
```suggestion
    if (!this.textLayer || !this.structTreeLayer) {
```

File: test/pdfs/bug1912001.pdf
Comment (Snuffleupagus) [inline test/pdfs/bug1912001.pdf:1]: This file doesn't *appear* to be used anywhere!?

File: web/pdf_page_view.js [hunk 9]
```diff
@@ -468,19 +472,15 @@ class PDFPageView {
...
       return;
     }
-    this.structTreeLayer ||= new StructTreeLayerBuilder();
 
-    const tree = await (!this.structTreeLayer.renderingDone
-      ? this.pdfPage.getStructTree()
-      : null);
-    const treeDom = this.structTreeLayer?.render(tree);
-    if (treeDom) {
+    const treeDom = await this.structTreeLayer.render();
```
Comment (Snuffleupagus) [inline web/pdf_page_view.js:476]: Please let's ensure that that the structTreeLayer is actually available when the method is invoked.
```suggestion
    const treeDom = await this.structTreeLayer?.render();
```

File: web/pdf_page_view.js [hunk 10]
```diff
@@ -468,19 +472,15 @@ class PDFPageView {
...
+    if (treeDom && this.canvas && treeDom.parentNode !== this.canvas) {
       // Pause translation when inserting the structTree in the DOM.
       this.l10n.pause();
-      this.canvas?.append(treeDom);
+      this.canvas.append(treeDom);
       this.l10n.resume();
     }
-    this.structTreeLayer?.show();
+    this.structTreeLayer.show();
```
Comment (Snuffleupagus) [inline web/pdf_page_view.js:483]: Please revert this change, to prevent any theoretical issues if the structTreeLayer was removed *while* the render-promise resolved.
```suggestion
    this.structTreeLayer?.show();
```

**Review by Snuffleupagus — APPROVED** [review]
r=me, thank you.
